### ISSUE TITLE
Recorder league display sorting incorrectly

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#648]
It's been pointed out that our recorder league page (https://www.brc.ac.uk/irecord/recorder-league) is not sorting properly. On my computer, when I first go to the page it displays by descending order of number of records. However, if I then try to sort the list on either of the numeric columns it is incorrectly sorting as if they were text rather than integers.

Also, it's been reported that even if you do sort by number of species recorded, you are still only seeing the top 100 record contributors, which is not the same thing as the top 100 species-diversity contributors.

**Comment by DavidRoy** [comment]
There are recurring forum posts on this.  @burkmarr could you have a quick look at this one when you start investigating ES reporting

**Comment by burkmarr** [comment]
I think that the sorting is not working because iForm needs to mark the two numeric columns in the table with the attribute ```data-type="numeric"```. @johnvanbreda is it possible form me to set up an 'Elasticsearch outputs (customisable)' form (i.e. replicate the recorder's league) on my local machine (alternatively on dev) that connects to the Elasticsearch index so that I can investigate a way to fix this?

**Comment by johnvanbreda** [comment]
@burkmarr do you have an instance of Elasticsearch on your machine populated with some data? Because of the proxying, the warehouse has to be on the same network as the Elasticsearch instance so if you are using a local warehouse you'll need a local ES instance. It would be possible I guess to point a local Drupal install at the live warehouse to get the proxying working though?

I think the sorting issue here is because of the type of aggregation (which I've called "simple") the way that the Elasticsearch request needs to be formatted to get sorting on different columns is a little fiddly. As it wasn't a priority at the time I just implemented a client side sort. Ideally though it would reissue a new request with a different sort.

I guess at some point you'll want to go through the process of installing a local Elastic stack and filling it with data.

**Comment by burkmarr** [comment]
Thanks @johnvanbreda - a local instance connecting to the live warehouse might be the way to go initially in order to get good ElasticSearch data into local forms and iForm code that I can modify and test things out.

**Comment by johnvanbreda** [comment]
Ok, so you just need a Drupal instance pointing to any website registered on the live warehouse. Then the Elasticsearch configuration can point you at the same alias as iRecord and you'll be good to go.

**Comment by burkmarr** [comment]
Can the website be running on localhost @johnvanbreda?

**Comment by johnvanbreda** [comment]
Yes, you can have localhost with Drupal on it, pointing to the live warehouse, then you just need the correct account details for the Elasticsearch alias.

**Comment by burkmarr** [comment]
Excellent - thanks @johnvanbreda.

**Comment by burkmarr** [comment]
@johnvanbreda - I created a new local D7 instance with iForm. When I configure iForm and try to point to the Warehouse, I get the message shown below. I have registered the website on the Warehouse and am using the correct ID and password. It works fine if I connect it to my local Warehouse. Any ideas about what might be going wrong?

![image](https://user-images.githubusercontent.com/9275178/72449521-b9d90200-37b0-11ea-848a-cc4c918eea1d.png)

**Comment by BirenRathod** [comment]
@burkmarr 
You need to link cacert.pem with PHP. you can configure pem file under curl sectio in php.ini file. 
You can download one from here -> https://curl.haxx.se/docs/caextract.html

**Comment by burkmarr** [comment]
Thanks @BirenRathod - that seems to have fixed that problem. But I have another one now - there's still some sort of problem connecting to ES. 

@johnvanbreda - I've simply reproduced the recorder's league page on my localhost with the same ES credentials. On the Warehouse I signed my local website up to the iRecord sharing agreement with the 'Receives data for reporting' option set. Can you think of any reason why this wouldn't work?

![image](https://user-images.githubusercontent.com/9275178/72512123-79769400-3843-11ea-8ed9-3ca008bff79c.png)

**Comment by BirenRathod** [comment]
@burkmarr 
No problem. That is specific to ES problem. Make sure credentials are same as iRecord live one.

**Comment by burkmarr** [comment]
@johnvanbreda. Sorry about my previous confusion. I've checked everything now and I am pointing at the live Warehouse. I know that I am connecting with it okay because I have set up a standard explore records page which delivers records from the Warehouse to my local website without problem. But my page reproducing the recorder's league - which I copied from iRecord with exactly the same ES settings - fails as shown below.

![image](https://user-images.githubusercontent.com/9275178/72516913-60bdac80-384a-11ea-805a-695f4f05323b.png)

**Comment by johnvanbreda** [comment]
@burkmarr on the page, there are some permissions settings which control access to your own records and all records separately. As this report aggregates other people's data, you need to make sure the all records permission is set appropriately. But that is perhaps not the issue here if you are logged in with admin rights.
If you look in the warehouse code for "Unable to authorise" you'll see there are 3 possibilities as to why this might occur. Maybe some temporary debug code on the warehouse could help?

**Comment by burkmarr** [comment]
@johnvanbreda - would it fail because my localhost doesn't have a security certificate?

**Comment by burkmarr** [comment]
@BirenRathod @johnvanbreda - I set an SSL certificate on my localhost so now requests are coming from https://localhost/drupal7w1b/content/recorder-league-table. However it still fails with the error message shown above (https://github.com/BiologicalRecordsCentre/iRecord/issues/648#issuecomment-575086514). 

But some logging I've temporarily put in ```modules\rest_api\controllers\services\rest.php``` seems to indicate that the requests from my localhost are still not considered to be secure. 

From function ```authenticate```, this line...
 ```kohana::log('alert', "isHttps: $this->isHttps");```
outputs this...
```2020-01-17 11:40:18 +00:00 --- alert: isHttps: ```

It only seems to be requests from my localhost that produce this. The logfile has plenty of successful requests - presumably from iRecord - which look like this:
```2020-01-17 11:40:22 +00:00 --- alert: isHttps: 1```

(I'm also puzzled as to why the the log doesn't read ```alert: isHttps: 0``` for the failures since it is set like this: ```$this->isHttps = isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on';```.)

Any help on this would be appreciated - I can't really begin to tackle the ES related work I've been assigned until I can establish a local Drupal/iForm instance that can consume the ES data.

**Comment by BirenRathod** [comment]
@burkmarr 
Basically your localhost should access with https, so your url should look like this -> https://localhost.
To assign SSL certificate for your localhost see this link -> 
https://stackoverflow.com/questions/8169999/how-can-i-create-a-self-signed-cert-for-localhost

**Comment by BirenRathod** [comment]
@burkmarr 
Also, you need to implement a value of $base_url in your settings.php in the Drupal folder. Add the https + your url.
And make sure iForm settings also have a warehouse URL with https.

**Comment by BirenRathod** [comment]
@burkmarr 
Have you taken backup from live iRecord website?

**Comment by BirenRathod** [comment]
Sorry I mean the data using Drupal's "backup and restore" link not the iForm folder.

**Comment by burkmarr** [comment]
@BirenRathod - I had already set up an SSL certificate (see start of https://github.com/BiologicalRecordsCentre/iRecord/issues/648#issuecomment-575600596).

I hadn't set $base_url in settings.php, so I did that (```$base_url = 'https://localhost/drupal7w1b';```), but it didn't make a difference.

I thought that I had specified a Warehouse with https, but when I checked, it was http. So I tried to change it to https, and got a bunch of errors. Then something made me think about the registration of the localhost website on the Warehouse and I realised that I hadn't changed it from http to https. So I changed that to https on the Warehouse and then I was able to change the Warehouse to https on the website!

And that did it - I can now consume the ES data from the website. So I now have a localhost Drupal/iForm that can consume ES data from Warehouse1 and I can start looking into this and other issues. Thanks for your help @BirenRathod and @johnvanbreda.

(@BirenRathod - I'm not sure what you mean about the backup. I have a local Drupal DB, but I'm viewing data from the iRecord website because my localhost website is registered on Warehouse1 and is signed up to the iRecord website agreement to report on iRecord data.)

**Comment by BirenRathod** [comment]
@burkmarr 
Well done you have resolved now that.
Ignore my query about backup, I asked to find out if the problem is related to that.

**Comment by burkmarr** [comment]
@kitenetter @DavidRoy - the first part of this is fixed now - i.e. the retrieved data now sorts correctly (numerically) on the _number of species_ and _number of records_ columns. This just required the addition of this attribute on the columns in :  `"data-type":"numeric"` in the data grid definition on the form (i.e. no code changes) - but it took me messing about with the code and putting in diagnostics on a local instance for me to find that out.

**Comment by burkmarr** [comment]
@johnvanbreda - the second part of this might be more problematic:

> Also, it's been reported that even if you do sort by number of species recorded, you are still only seeing the top 100 record contributors, which is not the same thing as the top 100 species-diversity contributors.

My plan was to replicate the form except modifying the ES query to bring back the recorders ordered by the aggregated species count. If I can do that, we can simply have the two pages and link them with some explanation and hyperlinks

But when I tried it I got weird results - ES didn't seem to bring back all the records. I tried on Kibana through the visualisation interface and got similar strange results, but on there I noticed some messages saying something like '3 of the 4 shards failed'. When I tried to run the query directly from the Kibana 'dev tools' interface, the ES index fell over.

Here's the current aggregation query that works:
```
  {
    "recorder_agg": {
      "terms": {
        "field": "event.recorded_by.keyword",
        "size": 100,
        "order": {
          "_count": "desc"
        }
      },
      "aggs": {
        "species_count": {
          "cardinality": {
            "field": "taxon.species_taxon_id"
          }
        }
      }
    }
  }
```
All I changed was this bit:
```
        "order": {
          "species_count": "desc"
        }
```

When I did this on Kibana from the visualisations interface and saw the '3 of 4 shards failed' toast popup, I inspected the JSON returned from the query and there was more detail in there. For the three shards that failed is said something about them requiring more memory than could be allocated.

**Comment by kitenetter** [comment]
@johnvanbreda to test on local ES instance, and look at options for enabling additional filtering (likely to be needed for other report pages in addition to the recorders' league).

**Comment by johnvanbreda** [comment]
@burkmarr I tested this on my local instance (approx 20K records) and it seemed fine. So, I tested on the live ES and sure enough, it crashed the service. I restarted it and everything was fine.

Now looking into any ways of changing the ES query to get around this. I guess it is simply forcing it to count across the entire index and breaking memory limits.

**Comment by burkmarr** [comment]
Thanks  @johnvanbreda.

**Comment by johnvanbreda** [comment]
I don't think there is a perfect solution here. The hashing used in the cardinality aggregation is memory intensive when there are thousands of buckets as there are here. One solution is to use precision_threshold to reduce the memory consumption on a less precise species count for sorting purposes, then include a 2nd aggregation with a more precise version for display purposes:
```
{
  "size": 0,
  "aggs": {
    "recorder_agg": {
      "terms": {
        "field": "event.recorded_by.keyword",
        "size": 100,
        "order": {"rough_species_count": "desc"},
        "min_doc_count": 10
      },
      "aggs": {
        "rough_species_count": {
          "cardinality": {
            "field": "taxon.species_taxon_id",
            "precision_threshold": 100
          }
        },
        "species_count": {
          "cardinality": {
            "field": "taxon.species_taxon_id"
          }
        }
      }
    }
  }
}
```

**Comment by burkmarr** [comment]
@johnvanbreda - for this to work without blowing the memory, is it necessary to do the ordering on 'rough_species_count'? I think I misunderstood how this works and I changed the ordering to 'species_count' and I think that broke ES again.

**Comment by johnvanbreda** [comment]
Correct. Rough_species_count is there only to provide an approximate search that uses lower memory consumption than species_count.

**Comment by burkmarr** [comment]
@johnvanbreda - Is this quite likely not to return the _real_ top 100 by species count?

**Comment by johnvanbreda** [comment]
It's a possibility - this is definitely a compromise. I suspect the order will only be wrong where 2 users are close. In which case, re-sorting the table by species count would fix it (as there is already code for a within table sort available). This would leave the small possibility that the person who just missed being included actually has a slightly higher species count than the 100th person.

**Comment by burkmarr** [comment]
@johnvanbreda - how can I specify the initial sort of the aggregated data on the species_count field? In the docs it says you can sort a source like this:

```@sort={"fieldName":"desc"}```

But it also says that is for non-aggregated output, so I guess it wouldn't work in this case? (I don't want to just try it in case ES falls over.)

**Comment by johnvanbreda** [comment]
If you look at the edit tab for the current recorder league table on iRecord you'll see that the aggregation contains the sort information embedded within it.

**Comment by burkmarr** [comment]
@johnvanbreda - I can only see the 'order' part, but changing ```"order": {"rough_species_count": "desc"}``` to ```"order": {"species_count": "desc"}``` caused the last crash.

**Comment by johnvanbreda** [comment]
You don't want to order on species_count, you want to order on rough_species_count. When elasticsearch does an order on an aggregation, it needs to load into memory a value for every aggregation bucket in the entire dataset. Doing this on recorders means hundreds of thousands of buckets. The amount of memory required to do this depends on the accuracy of the hashing algorithm used, defined by the precision_threshold. So we use rough_species_count to put the data into the approximately correct order using a low-precision approach, then use a more precise version of the same field to display the value. This latter more precise version only gets calculated for the current page of data so doesn't use much memory.

Does that make sense?

**Comment by burkmarr** [comment]
@johnvanbreda - Yes it does, but I don't think I described my problem properly. The grid is initially sorted on rough_species_count, but there are places where this is not the right order for species_count. Users can use the interactive sorting on the field to fix it, but I wanted to avoid the need to do that if possible. The page is here: https://www.brc.ac.uk/irecord/recorder-league-sp. You can see, for example, that the fifth and sixth rows are in the wrong order.

**Comment by johnvanbreda** [comment]
Ok, but the solution to this problem is not to ask ES to sort by species_count, since that breaks ES. I think there are 2 possible solutions:
1) Automate the click on the interactive sorting on the field so the user doesn't have to do this, as soon as the response comes back (or just fire the sort code directly).
2) Do 2 hits on the ES service. First, select, say, the top 150 recorders ordered by rough_species_count. Then do a second hit ordered on species_count, but this time with an outer filter to limit to the top 150 recorders returned by the first query.

Idea 2 sounds more elegant and powerful but more work - basically a way of daisychaining sources so each is filtered by the results of its predecessor. I couldn't see anything built into the ES API for this. Our config for this might look like the following:

```
[source]
@id=roughTop150
@aggregation= ...

[source]
@id=top100
@aggregation= ...
@filterBoolClauses=<!--{
  "must":[
    {"query_type":"terms", "field":"event.recorded_by.keyword", "valueFromSource": {
      "source": "roughTop150",
      "field": "recorder_agg.value"
    }
  ]
}-->
```

That's all very rough, just to give the idea.

**Comment by burkmarr** [comment]
Thanks John - I used the first method. I put some code in \sites\default\files\indicia\js\node.7916.js to automate the sort click after the table is populated for the first time. I'll bear the second option in mind as I learn more about ES.

**Comment by johnvanbreda** [comment]
@burkmarr - I've added a new method of handling tabular data containing aggregations like this, which supports sorting in a single table including the sorting on an approximation aggregation to avoid resource usage issues. See https://indicia-docs.readthedocs.io/en/latest/site-building/iform/helpers/elasticsearch-report-helper.html#automatically-generating-aggregations

**Comment by burkmarr** [comment]
@johnvanbreda - I used the example documented in the docs on live iRecord and it was okay, but when I repeated on dev - which I'd updated with latest develop branch (both warehouse and iRecord) - I got this error:

![image](https://user-images.githubusercontent.com/9275178/79793750-ffc1db00-8348-11ea-883c-99fb0cda8b6a.png)

Does that ring any bells with you? I've checked that all the relevant ES fields are included in the index and they are.

**Comment by BirenRathod** [comment]
@burkmarr 
There is a fair chance that that field is not exists in the table.

**Comment by burkmarr** [comment]
@Biren - all the fields are in the ES index, so I don't think it can be that.

**Comment by BirenRathod** [comment]
@burkmarr are they exists in the table also?

**Comment by johnvanbreda** [comment]
The develop and media branches didn't have the latest code- I just checked out develop and git pulled which has fixed this message. Can I look at this again though before you go further - I suspect that sorting by record count descending might cause memory issues in ES.

**Comment by burkmarr** [comment]
@johnvanbreda - I was just looking and think I found the problem: it was faling because the aggregation was looking for field 'id' - I changed it to 'event.event_id' and it worked. 

This failed:
```
  "aggs": {
    "occs_count": {
      "cardinality": {
        "field": "id"
      }
    },
```
This worked:
```
"aggs": {
    "occs_count": {
      "cardinality": {
        "field": "event.event_id"
      }
    },
```
Did you do something to enable you to use id as an alias for event.event_id in this situation or could that be an ES v6 vs ES v7 thing?

**Comment by burkmarr** [comment]
@johnvanbreda - sorry I missed your previous message. I looks as though you found the real problem.

**Comment by burkmarr** [comment]
@johnvanbreda. Can I just check a git question with you? I thought I'd pulled the latest version by doing:
```
git pull --recurse-submodules
```
What am I misunderstanding about how that works?

**Comment by johnvanbreda** [comment]
@burkmarr pulling with a --recurse-submodules option should work I think, but only if I've been fastidious enough to commit to the develop branch of iform the tree positions of the submodules. I tend to only bother doing this when releasing on the master branch so the develop branch of iform may point to an old head position of the submodules. Hence the need to cd into them and git checkout develop then pull each submodule separately.

I think there was a muddle up between id and event_id in my demo code so good spot.

I've now committed some more changes to the develop branch:
* Made the way the @sort option works consistent for normal data sources and autoAggregationTables (i.e. no need for an `order` subkey).
* Enabled the use of doc_count as if it were an aggregated field. That saves adding occs_count in this case and seems to work better for sorts. Sorting on occs_count breaks the memory limit unless a high precision_threshold is set.

I've updated the example on the devserver, as well as the docs (though they haven't been rebuilt yet).

**Comment by burkmarr** [comment]
Great thanks John - I'll will remember that now re pulling. Thanks also for the info on the updated aggregations stuff - I will look at that later.

**Comment by burkmarr** [comment]
@johnvanbreda - I just looked again at http://test.brc.ac.uk/irecord_dev/node/7620 and I can see the changes you made. But when I attempt to sort on the species I get an ES error.

**Comment by DavidRoy** [comment]
@burkmarr seems odd that the record count on the table is > 4 million.  Do we really have that many 'users'?

**Comment by burkmarr** [comment]
@DavidRoy - that's just a test page listing samples not recorders, so its 4 million samples.

**Comment by DavidRoy** [comment]
@burkmarr that makes sense.  Still an impressive number!

**Comment by johnvanbreda** [comment]
@burkmarr the error when sorting is a "data too large" error which I found in the Elasticsearch log file. Also in the warehouse log file it logs the query and response when an ES error occurs. I checked the query sent and it was correctly using the column with the precision_threshold set to reduce the memory consumption but I guess it is still too much when aggregating several million buckets. Maybe if you switch this test over to using a row per recorder as originally intended the problem will go away?

**Comment by burkmarr** [comment]
@johnvanbreda - I've converted the example referred to above (http://test.brc.ac.uk/irecord_dev/node/7620/edit) to group by recorders instead of samples, but I have met a problem with retrieving recorder names. The page as it stands attempts to retrieve 'event.recorded_by.keyword' as per the current recorders league on the live site. But although the query appears to work, the recorder's names are blank. It made me think about that field: event.recorded_by.keyword. What is the '.keyword' suffix? In Kibana, the field is addressed just by 'event.recorded_by' but if I just use that in this form, it falls over. (Note that the general structure of everything looks okay because if I just swap out 'event.recorded_by.keyword' for 'event.event_id', it aggregates by event_id okay.)

**Comment by johnvanbreda** [comment]
In Elasticsearch, string data can be stored in a text datatype or a keyword datatype. The former is full-text indexed so good for fuzzy searches. The latter is stored as a serious of keywords so good for exact searches or aggregations. Sometimes we store both - which Elasticsearch does by creating a version of the field with a .keyword suffix. 

When requesting the data columns, Indicia uses source filtering to limit the fields in the response. It automatically includes your unique field in this BUT Elasticsearch's source filtering want's the base name of the field without the .keyword suffix otherwise the fieldname is ignored. So, my code to automatically get your unique field in the response does not work in this case. 

I've applied a quick fix by including event.recorded_by in the fields array in the source. I've also committed a change so this is handled automatically: https://github.com/Indicia-Team/media/commit/e89804f979bdaa1b62ed41149743c6374156aaba

**Comment by burkmarr** [comment]
@johnvanbreda - I've tried this with the new ES client-side code in the develop branch (on my local instance linked to dev warehouse/ES) and it works beautifully from the example given in the documentation (https://indicia-docs.readthedocs.io/en/latest/site-building/iform/helpers/elasticsearch-report-helper.html#typical-configuration-examples) with one change which is changing compositeAggregation for termAggregation. I guess that is a mistake in the docs? I'm not completely clear on the difference between these two aggregation modes.

**Comment by burkmarr** [comment]
(@johnvanbreda There are also a couple of references to autoAggregationTable in the documentation which I think are no longer right? Other than that I worked through the documentation yester working many of the examples and experimenting and it is great.)

**Comment by burkmarr** [comment]
I will implement the new code on the live recorder's page when the new code is merged into the master branch and deployed on iRecord live.

**Comment by johnvanbreda** [comment]
@burkmarr 
* I've updated the documentation to correct the references to autoAggregationTables which are no longer in the code. 
* The example given using a compositeAggregation should actually have been a termAggregation because it includes a sortAggregation which is only available in termAggegration mode. I've corrected this in the docs. However the problem you encountered using compositeAggregation was probably the result of a bug in ElasticsearchReportHelper.php - I've committed a fix to develop.
* In the documentation for the mode option, I've added information on the difference between composite and term aggregation modes. Both have limitations unfortunately.
